### PR TITLE
UCP/EP/PROTO: Print endpoint locality with protov2 select info

### PR DESCRIPTION
**Pull request by yosefe**
## Why
Improve protocol information output by printing endpoint locality information

Replaces #7902 

Example output:
```
+----------------+--------------------------------------------------------------------------------------------+
| mpi self cfg#0 | tagged message by ucp_tag_send*() from host memory                                         |
+----------------+-------------------------------------------+------------------------------------------------+
|        0..8184 | eager short                               | self/memory                                    |
|    8185..37142 | eager copy-in copy-out                    | self/memory                                    |
|     37143..inf | (?) rendezvous zero-copy read from remote | 68% on knem/memory and 32% on rc_mlx5/mlx5_0:1 |
+----------------+-------------------------------------------+------------------------------------------------+
+----------------------+--------------------------------------------------------------------------------------------+
| mpi intra-node cfg#1 | tagged message by ucp_tag_send*() from host memory                                         |
+----------------------+-------------------------------------------+------------------------------------------------+
|                0..92 | eager short                               | sysv/memory                                    |
|             93..8248 | eager copy-in copy-out                    | sysv/memory                                    |
|         8249..171045 | eager copy-in copy-out                    | sysv/memory                                    |
|          171046..inf | (?) rendezvous zero-copy read from remote | 68% on knem/memory and 32% on rc_mlx5/mlx5_0:1 |
+----------------------+-------------------------------------------+------------------------------------------------+
```

### CODE REVIEW

**Review by brminich — DISMISSED** [review]

**Comment by yosefe** [comment]
@brminich can you pls take a look?

File: src/ucp/wireup/select.c [hunk 1]
```diff
@@ -2043,7 +2043,70 @@ static unsigned ucp_wireup_default_path_index(unsigned path_index)
...
+                key->flags |= UCP_EP_CONFIG_KEY_FLAG_INTRA_NODE;
+                ucs_free(dev_addr);
+                return UCS_OK;
+            }
+        }
+
+        ucs_free(dev_addr);
+    }
+
+    return UCS_OK;
```
Comment (brminich) [inline src/ucp/wireup/select.c:2106]:
```suggestion
        status = uct_iface_get_device_address(wiface->iface, dev_addr);
        if (status != UCS_OK) {
            goto out;
        }

        ucp_unpacked_address_for_each(ae, select_params->address) {
            if ((wiface->attr.device_addr_len == ae->dev_addr_len) &&
                !memcmp(ae->dev_addr, dev_addr, ae->dev_addr_len)) {
                key->flags |= UCP_EP_CONFIG_KEY_FLAG_INTRA_NODE;
                goto out;
            }
        }


    }
out:
    ucs_free(dev_addr);
    return status;
```

File: src/ucp/wireup/select.c [hunk 2]
```diff
@@ -2174,6 +2237,9 @@ ucp_wireup_construct_lanes(const ucp_wireup_select_params_t *select_params,
     /* use AM lane first for eager AM transport to simplify processing single/middle
      * msg packets */
     key->am_bw_lanes[0] = key->am_lane;
+
+    /* Set locality flags */
```
Comment (brminich) [inline src/ucp/wireup/select.c:2241]: imo, this is excessive comment - the same info is provided by func name

File: src/ucp/wireup/select.c [hunk 3]
```diff
@@ -2214,7 +2280,11 @@ ucp_wireup_select_lanes(ucp_ep_h ep, unsigned ep_init_flags,
     }
 
 out:
-    ucp_wireup_construct_lanes(&select_params, &select_ctx, addr_indices, key);
+    status = ucp_wireup_construct_lanes(&select_params, &select_ctx,
+                                        addr_indices, key);
+    if (status != UCS_OK) {
+        return status;
+    }
```
Comment (brminich) [inline src/ucp/wireup/select.c:2287]: is it unrelated bug fix?

File: src/uct/tcp/tcp_iface.c [hunk 4]
```diff
@@ -128,6 +128,7 @@ static ucs_status_t uct_tcp_iface_get_device_address(uct_iface_h tl_iface,
 
     if (ucs_sockaddr_is_inaddr_loopback(saddr)) {
         dev_addr->flags |= UCT_TCP_DEVICE_ADDR_FLAG_LOOPBACK;
+        memset(pack_ptr, 0, sizeof(uct_iface_local_addr_ns_t));
```
Comment (brminich) [inline src/uct/tcp/tcp_iface.c:131]: is it some bug fix?

File: src/ucp/wireup/select.c (hunk 1 above)
Comment (yosefe) [reply src/ucp/wireup/select.c:2106]: so dev_addr can leak if the loop is run more than once.. would need separate err flow and good flow (since need to release dev_addr inside the loop)

File: src/ucp/wireup/select.c (hunk 3 above)
Comment (yosefe) [reply src/ucp/wireup/select.c:2287]: one of the gtests replaces get_device_address by a stub that returns error and expects wireup to fail if this stub is called at least once. so this change makes sure we fail whenever get_device_address fails from the context of the new locality code

File: src/uct/tcp/tcp_iface.c (hunk 4 above)
Comment (yosefe) [reply src/uct/tcp/tcp_iface.c:131]: yes, it's an uninitialized memory issue found with valgrind, probably because we are calling get_device_address from another flow with this PR

File: src/ucp/wireup/select.c (hunk 1 above)
Comment (brminich) [reply src/ucp/wireup/select.c:2106]: yeah, missed it

**Review by brminich — APPROVED** [review]

**Review by brminich — APPROVED** [review]
